Address many deprecation warnings

mwoodiupui opened the pull request:
## References
* Fixes [#2956](https://github.com/DSpace/DSpace/issues)

## Description
Address many deprecation warnings.

## Instructions for Reviewers
It may be easiest to review each commit separately, since I've tried to group all work by the deprecated code it addresses.

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
Remove long-deprecated ConfigurationManager.

mwoodiupui committed:
Replace deprecated Spring @Required with @Autowired(required=true)

mwoodiupui committed:
Replace deprecated Commons Collections MultivalueMap.

mwoodiupui committed:
Replace deprecated Commons CLI PosixParser and OptionBuilder.

mwoodiupui committed:
Replace deprecated Xerces XML serializer with one from org.w3c.dom.

mwoodiupui committed:
Replace deprecated lang3.CharEncoding with nio.StandardCharsets.

mwoodiupui committed:
Remove deprecated getMetadata(Context, String) from CommunityService, CollectionService.

mwoodiupui committed:
Better fix for configuration issue in c463b632dd24e0c017d88b36b2b6fa0f13d95fb1

Other code now depends on the 'id' of the curation script configuration bean,
so it can't be renamed.  Instead, remove the bean post processing code that
pokes 'dspace.cfg' properties into a bean which are prefixed with the bean's
'id', since nobody can remember why we do that and it causes subtle magical
problems.  It may be for Guice, which is no longer an option for DSpace.

tdonohue commented:
@mwoodiupui : Just a quick note...it looks like this has a single test failure in `DiscoveryRestControllerIT`.  There's also a minor, new alert that LGTM points out above in `S3BitStoreService`.  Once those are cleaned up though, I'll gladly give this a review.

mwoodiupui committed:
I missed a lot of deprecated PosixParser and OptionBuilder in the clutter of warnings (see?).

mwoodiupui committed:
Missed one more PosixParser.

mwoodiupui committed:
Fix broken test by catching an unexpected exception.

There must be an unchecked exception bubbling up through SolrServiceImpl#search.  Catching Exception (as before this PR) fixes the test failure.

mwoodiupui committed:
Replace deprecated constructors of boxed primitive types. [#2956]

mwoodiupui committed:
Replace deprecated Class.newInstance(). [#2956]

mwoodiupui committed:
Fix anything that is documented @deprecated but not annotated @Deprecated. [#2956]

mwoodiupui committed:
Replace various deprecated PDFBox methods. [#2956]

mwoodiupui committed:
Remove long-deprecated un-localized methods from LocationUtils. [#2956]

mwoodiupui committed:
Fix LGTM alert:  pointless boxing.

tdonohue commented:
@mwoodiupui : just a quick FYI...this needs rebasing on `main` as it has merge conflicts & also looks to have previously hit the Travis 50min build time limit

mwoodiupui committed:
I missed one LocationUtils call.

mwoodiupui committed:
Remove deprecated setMetadata of CommunityService, CollectionService #2956

This introduces a new class:  a holder for metadata field names.

mwoodiupui committed:
Found a few more deprecated PosixParser, GnuParser, OptionBuilder. #2956

mwoodiupui committed:
Replace deprecated org.apache.http.impl.client.DefaultHttpClient [#2956]

mwoodiupui committed:
Replace miscellaneous rarely-used deprecateds. [#2956]

tdonohue commented:
@mwoodiupui : Any plans to get this updated in the near(ish) future?  I'd love to include this in 7.0 Beta 5 (likely due in Dec or very early Jan), if it can get updated as the deprecation warnings are also a thorn in my side.  If you aren't able to get to this soon, let me know & maybe I can find time in the coming month to help move this work forward.

mwoodiupui committed:
Replace deprecated setMetadata() that uses removed field names. #2956

Once upon a time, DSOs had dedicated fields for some metadata.  These were
moved into the Metadatavalue table long ago, but there were still methods
which used the old field names instead of proper MetadataField triplets.

mwoodiupui committed:
Satisfy checkstyle.

mwoodiupui committed:
Replace deprecated icu4j Normalizer with Normalizer2. #2956

mwoodiupui commented:
There are a few deprecations which are non-trivial and probably should be separate issues:

- org.dspace.core.Context implements #finalize().  Current best practice seems to be to use a Cleaner instead.
- org.dspace.eperson.service.EPersonService and .GroupService have #findAll methods which are deprecated in favor of paginated versions.  In the cases that I've examined, pagination would just clutter the caller's code because it simply goes straight down the List in one swoop.  May I suggest a findAll variant that returns an Iterable and manages the pagination internally.
- org.dspace.browse.BrowseIndex has several deprecated #getTableName methods which seem to be used only in a single database migration.  I don't know why they are deprecated.  I think that this needs discussion.

mwoodiupui committed:
Replace Class.getInstance(). #2956

mwoodiupui committed:
Replace deprecated JUnit and Mockito statics. #2956

mwoodiupui committed:
Satisfy Checkstyle.  Small tidy-ups.  #2956

mwoodiupui commented:
I'm inclined to say this is finished, even though there are a few deprecations left.  As noted above, they are more intrusive or require discussion.

abollini reviewed (commented):
thanks @mwoodiupui for this massive effort!
I have built the PR locally and run a couple of the touched Cli, everything work as expected. I have inspected the first 150 changed files (me and @tdonohue are trying to split the review work between us). All the changes look good, I have only added few inline comments.

Can you please solve the merge conflict?

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2956

mwoodiupui committed:
Address review comments.  #2956

mwoodiupui commented on dspace-api/src/main/java/org/dspace/app/util/GoogleMetadata.java:
```diff
@@ -180,9 +186,7 @@
-        if (log.isDebugEnabled()) {
-            logConfiguration();
-        }
+        logConfiguration();
```
Normally I would not use #isDebugEnabled but, since the code is iterating over a HashMap and logging its content, the added efficiency is probably worthwhile.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/app/statistics/HTMLReport.java:
```diff
@@ -59,7 +60,9 @@
      */
-    private String output = ConfigurationManager.getProperty("dspace.dir") +
+    ConfigurationService configurationService
+            = DSpaceServicesFactory.getInstance().getConfigurationService();
```
I've moved the initialization inside the formerly empty constructor.

I believe that this feature is due to be removed anyway after 7.0, so it is probably not worth the effort to figure out what could be removed now.

mwoodiupui commented on dspace-api/pom.xml:
```diff
@@ -496,7 +496,7 @@
         <dependency>
             <groupId>org.hamcrest</groupId>
-            <artifactId>hamcrest-core</artifactId>
+            <artifactId>hamcrest-all</artifactId>
```
We should consider it.  I can add that to this PR, but perhaps it should be a separate issue since it touches a lot of tests.

mwoodiupui commented:
Cleaning up all of these warnings has exposed more complaints from `errorprone`, so the build output is still huge.  I've begun looking at those.

tdonohue reviewed (commented):
@mwoodiupui : This looks great overall. I've done a code review of all of it, concentrating more on the classes which @abollini didn't review as well.  I found one minor issue through 9 different classes (see inline below)...it's not a bug though, just an additional refactor that could be done to simplify the code in many areas.

In any case, I'd recommend we find a way to wrap up this PR soon.  It's getting overly large, and it looks good "as-is", so this might be a good point to do some "final, brief cleanup"....fix the merge conflicts, rebase on latest `main` (to pick up the new GitHub CI builds), etc.  That way we can work to get this merged in the nearish future, and then you could follow it up with additional cleanup in a second PR.  Thanks!

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/packager/METSManifest.java:
```diff
@@ -190,21 +193,20 @@ public InputStream getInputStream(Element mdRef)
-        Enumeration<String> pe = (Enumeration<String>) ConfigurationManager.propertyNames();
-        StringBuffer result = new StringBuffer();
-        while (pe.hasMoreElements()) {
+        List<String> configKeys = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/crosswalk/XSLTDisseminationCrosswalk.java:
```diff
@@ -146,18 +149,17 @@ private void init()
         //  crosswalk.diss.{PLUGIN_NAME}.namespace.{PREFIX} = {URI}
         String nsPrefix = prefix + "namespace.";
-        Enumeration<String> pe = (Enumeration<String>) ConfigurationManager.propertyNames();
+        List<String> configKeys = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/crosswalk/XSLTCrosswalk.java:
```diff
@@ -117,11 +116,12 @@
-        Enumeration<String> pe = (Enumeration<String>) ConfigurationManager.propertyNames();
+        ConfigurationService configurationService
+                = DSpaceServicesFactory.getInstance().getConfigurationService();
+        List<String> configKeys = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/crosswalk/QDCCrosswalk.java:
```diff
@@ -218,32 +221,31 @@ private void init()
-        List<Namespace> nsList = new ArrayList<Namespace>();
-        Enumeration<String> pe = (Enumeration<String>) ConfigurationManager.propertyNames();
+        List<Namespace> nsList = new ArrayList<>();
+        List<String> configKeys = configurationService.getPropertyKeys();
```
Agreed.

tdonohue replied:
@mwoodiupui : Looks like you overlooked this one on accident.  Should be `configurationService.getPropertyKeys(propname)`

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/crosswalk/QDCCrosswalk.java:
```diff
@@ -137,11 +141,10 @@
-        List<String> aliasList = new ArrayList<String>();
-        Enumeration<String> pe = (Enumeration<String>) ConfigurationManager.propertyNames();
+        List<String> aliasList = new ArrayList<>();
+        List<String> configKeys = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/authority/MetadataAuthorityServiceImpl.java:
```diff
@@ -81,13 +84,12 @@ protected MetadataAuthorityServiceImpl() {
-            isAuthorityRequired = new HashMap<String, Boolean>();
-            Enumeration pn = ConfigurationManager.propertyNames();
+            isAuthorityRequired = new HashMap<>();
+            List<String> pn = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/crosswalk/MODSDisseminationCrosswalk.java:
```diff
@@ -107,10 +110,9 @@
-        while (pe.hasMoreElements()) {
-            String key = pe.nextElement();
+        List<String> aliasList = new ArrayList<>();
+        List<String> pe = configurationService.getPropertyKeys();
```
Agreed.

mwoodiupui commented on dspace-swordv2/src/main/java/org/dspace/sword2/AbstractSimpleDC.java:
```diff
@@ -27,16 +28,18 @@
                 String keyString = (String) key;
-                if (keyString.startsWith("simpledc.")) {
-                    String k = keyString.substring("simpledc.".length());
+                if (keyString.startsWith("swordv2-server.simpledc.")) {
```
Agreed.

mwoodiupui commented on dspace-swordv2/src/main/java/org/dspace/sword2/AbstractSimpleDC.java:
```diff
@@ -45,12 +48,11 @@ protected void loadMetadataMaps() {
                 String keyString = (String) key;
-                if (keyString.startsWith("atom.")) {
-                    String k = keyString.substring("atom.".length());
+                if (keyString.startsWith("swordv2-server.atom.")) {
```
Agreed.

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2956

mwoodiupui committed:
Address review comments.  #2956

mwoodiupui committed:
Satisfy checkstyle.  #2956

tdonohue reviewed (commented):
@mwoodiupui : Looking good now!  I did find one prior request that you accidentally overlooked (see below).  In the meantime though, I'll also download this PR and give it a test to be sure everything still seems to work as expected.

**UPDATE:** I've tested this now and did some quick "sanity checks" of both REST APIs, OAI, and both SWORD interfaces....everything seems to still be working best I can tell.  So, once this one last change is made, I'm +1 this PR.

mwoodiupui committed:
Address one more review comment.  New unit test class to test it.

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2956

mwoodiupui committed:
Merge branch 'main' into #2956

tdonohue reviewed (approved):
👍 This looks great to me now, @mwoodiupui !  Thanks for your hard work on fixing these annoying deprecation warnings.  As noted, I've already tested this and found no issues.

@abollini : have you had a chance to glance back at this?  Anything else you see that needs addressing?  It looks to me like your prior feedback was all addressed or answered (and I'd like to recommend we merge this soonish, as the PR is already very large).

abollini reviewed (approved):
thanks again @mwoodiupui for this huge effort, my previous feedbacks as been addressed and the further improvements agreed with Tim looks very good to me.
